src/js/tabs/osd.js

```diff
@@ -2184,33 +2193,44 @@ TABS.osd.initialize = function (callback) {
                             }
                         } else {
                             var arrayElements = field.preview;
-
-                            // The array can have negative and positive positions, search limits...
-                            var limits = OSD.searchLimitsElement(arrayElements);
-
-                            var offsetX = 0;
-                            var offsetY = 0;
-                            if (limits.minX < 0) {
-                                offsetX = -limits.minX;
-                            }
-                            if (limits.minY < 0) {
-                                offsetY = -limits.minY;
-                            }
-
                             for (var i = 0; i < arrayElements.length; i++) {
-
-                                // Add the character to the preview
                                 var element = arrayElements[i];
-                                var charCode = element.sym;
-                                OSD.drawByOrder(selectedPosition + element.x + element.y * FONT.constants.SIZES.LINE, field, charCode, element.x, element.y);
-
-                                // Image used when "dragging" the element
-                                if (field.positionable) {
-                                    var img = new Image();
-                                    img.src = FONT.draw(charCode);
-                                    ctx.drawImage(img, (element.x + offsetX) * 12, (element.y + offsetY) * 18);
-                                }
+				//Add string to the preview.
+	                        if (element.constructor == String) { 
```

Comment:
Not sure if adding this is actually the right thing to do - this adds complexity, and the same effect can be achieved easily by returning an array of objects, like it is done in https://github.com/betaflight/betaflight-configurator/blob/4dcc9105d386a20173d0b53abb8e925ad280b7d2/src/js/tabs/osd.js#L474-L509.

Also, this does not check to make sure the element can fit onto the screen.

Reply:
This was what I initially looked at but this only does one char per element as far I can see. I have thought of iterating through each strings and create individual element for each character. However this becomes very messy.  I am new to this so I may have missed something. 

RE making sure the element can fit, are you referring to this code below which re-position the element when it is dropped onto the canvas? This can be expanded. 

`        if (!display_item.ignoreSize) {
            if (display_item.preview.constructor !== Array) {
                // Standard preview, string type
                var overflows_line = FONT.constants.SIZES.LINE - ((position % FONT.constants.SIZES.LINE) + display_item.preview.length);
                if (overflows_line < 0) {
                    position += overflows_line;
                }
            } else {
                // Advanced preview, array type
                var arrayElements = display_item.preview;
                var limits = OSD.searchLimitsElement(arrayElements);

                var selectedPositionX = position % FONT.constants.SIZES.LINE;
                var selectedPositionY = Math.trunc(position / FONT.constants.SIZES.LINE);

                if ((limits.minX < 0) && ((selectedPositionX + limits.minX) < 0)) {
                    position += Math.abs(selectedPositionX + limits.minX);
                } else if ((limits.maxX > 0) && ((selectedPositionX + limits.maxX) >= FONT.constants.SIZES.LINE)) {
                    position -= (selectedPositionX + limits.maxX + 1) - FONT.constants.SIZES.LINE;
                }
                if ((limits.minY < 0) && ((selectedPositionY + limits.minY) < 0)) {
                    position += Math.abs(selectedPositionY + limits.minY) * FONT.constants.SIZES.LINE;
                } else if ((limits.maxY > 0) && ((selectedPositionY + limits.maxY) >= OSD.data.display_size.y)) {
                    position -= (selectedPositionY + limits.maxY - OSD.data.display_size.y + 1) * FONT.constants.SIZES.LINE;
                }
            }
        }

`

Reply:
Good point, I thought the array of objects could be used for strings as well. If not, extending that, and allowing the strings to be anchored to an arbitrary point makes more sense in my opinion.
And yes - we want to prevent users from placing elements off screen.